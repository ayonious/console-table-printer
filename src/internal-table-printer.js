const { COLOR } = require('./table-constants');
const { ColoredConsoleLine } = require('./colored-console-line');
const { textWithPadding,
    printTableHorizontalBorders,
    createRow,
    createHeaderAsRow,
    findMaxLenOfColumn
} = require('./table-helpers');


function prepareLineAndPrint(tableStyle, columns, row) {
    let line = new ColoredConsoleLine();
    line.addWithColor(COLOR.white, tableStyle.vertical);
    for (let column of columns) {
        line.addWithColor(COLOR.white, ' ');
        line.addWithColor(row.color, textWithPadding(`${row.text[column.name] || ''}`, column.alignment, column.max_ln));
        line.addWithColor(COLOR.white, ' ' + tableStyle.vertical);
    }
    return line.printConsole();
}

// ║ 1     ║     I would like some red wine please ║ 10.212 ║
function printRow(table, row) {
    let ret = [];
    ret.push(prepareLineAndPrint(table.tableStyle, table.columns, row));
    return ret;
}


/* 
 ╔═══════╦═══════════════════════════════════════╦════════╗
 ║ index ║                                  text ║  value ║
 ╟═══════╬═══════════════════════════════════════╬════════╢
*/
function printTableHeaders(table) {
    let ret = [];

    // ╔═══════╦═══════════════════════════════════════╦════════╗
    ret.push(printTableHorizontalBorders(table.tableStyle.headerTop,
        table.columns.map(m => m.max_ln)));

    // ║ index ║                                  text ║  value ║
    let row = createHeaderAsRow(createRow, table.columns);
    ret.push(prepareLineAndPrint(table.tableStyle, table.columns, row));

    // ╟═══════╬═══════════════════════════════════════╬════════╢
    ret.push(printTableHorizontalBorders(table.tableStyle.headerBottom,
        table.columns.map(m => m.max_ln)));

    return ret;
}

function printTableEnding(table) {
    let ret = [];
    // ╚═══════╩═══════════════════════════════════════╩════════╝
    ret.push(printTableHorizontalBorders(table.tableStyle.tableBottom,
        table.columns.map(m => m.max_ln)));
    return ret;
}

function calculateColumnProperty(table) {
    for (let column of table.columns) {
        column.max_ln = findMaxLenOfColumn(column, table.rows);
    }
}

function printTableTest(table) {
    calculateColumnProperty(table);
    let ret = [];
    printTableHeaders(table).forEach(row => ret.push(row));
    for (let row of table.rows) {
        printRow(table, row).forEach(row => ret.push(row));
    }
    printTableEnding(table).forEach(row => ret.push(row));
    return ret;
}

function printTable(table) {
    printTableTest(table);
}

function printSimpleTableTest(rows) {
    const { TableInternal } = require('./internal-table');
    let table = new TableInternal();
    table.addRows(rows);
    return printTableTest(table);
}

function printSimpleTable(rows) {
    printSimpleTableTest(rows);
}

module.exports = {
    printTable,
    printSimpleTable,

    // test functions
    printSimpleTableTest
}

