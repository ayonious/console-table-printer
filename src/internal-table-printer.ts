import { COLOR, TABLE_STYLE_DETAILS } from "./table-constants";
import { ColoredConsoleLine } from "./colored-console-line";
import {
  textWithPadding,
  printTableHorizontalBorders,
  createRow,
  createHeaderAsRow,
  findMaxLenOfColumn,
  Row,
  Column
} from "./table-helpers";
import { TableInternal } from "./internal-table";

function prepareLineAndPrint(
  tableStyle: TABLE_STYLE_DETAILS,
  columns: Column[],
  row: Row
): string {
  let line = new ColoredConsoleLine();
  line.addWithColor(COLOR.white, tableStyle.vertical);
  for (let column of columns) {
    line.addWithColor(COLOR.white, " ");
    line.addWithColor(
      row.color,
      textWithPadding(
        `${row.text[column.name] || ""}`,
        column.alignment,
        column.max_ln
      )
    );
    line.addWithColor(COLOR.white, " " + tableStyle.vertical);
  }
  return line.printConsole();
}

// ║ 1     ║     I would like some red wine please ║ 10.212 ║
function printRow(table: TableInternal, row: Row): string[] {
  let ret: string[] = [];
  ret.push(prepareLineAndPrint(table.tableStyle, table.columns, row));
  return ret;
}

/* 
 ╔═══════╦═══════════════════════════════════════╦════════╗
 ║ index ║                                  text ║  value ║
 ╟═══════╬═══════════════════════════════════════╬════════╢
*/
function printTableHeaders(table: TableInternal): string[] {
  let ret: string[] = [];

  // ╔═══════╦═══════════════════════════════════════╦════════╗
  ret.push(
    printTableHorizontalBorders(
      table.tableStyle.headerTop,
      table.columns.map(m => m.max_ln)
    )
  );

  // ║ index ║                                  text ║  value ║
  let row = createHeaderAsRow(createRow, table.columns);
  ret.push(prepareLineAndPrint(table.tableStyle, table.columns, row));

  // ╟═══════╬═══════════════════════════════════════╬════════╢
  ret.push(
    printTableHorizontalBorders(
      table.tableStyle.headerBottom,
      table.columns.map(m => m.max_ln)
    )
  );

  return ret;
}

function printTableEnding(table: TableInternal): string[] {
  let ret: string[] = [];
  // ╚═══════╩═══════════════════════════════════════╩════════╝
  ret.push(
    printTableHorizontalBorders(
      table.tableStyle.tableBottom,
      table.columns.map(m => m.max_ln)
    )
  );
  return ret;
}

function calculateColumnProperty(table: TableInternal) {
  for (let column of table.columns) {
    column.max_ln = findMaxLenOfColumn(column, table.rows);
  }
}

export function printTableTest(table: TableInternal): string[] {
  calculateColumnProperty(table);
  let ret: string[] = [];
  printTableHeaders(table).forEach(row => ret.push(row));
  for (let row of table.rows) {
    printRow(table, row).forEach(row => ret.push(row));
  }
  printTableEnding(table).forEach(row => ret.push(row));
  return ret;
}

export function printTable(table: TableInternal) {
  printTableTest(table);
}

export function printSimpleTableTest(rows: any[]) {
  let table = new TableInternal();
  table.addRows(rows);
  return printTableTest(table);
}

export function printSimpleTable(rows: any[]) {
  printSimpleTableTest(rows);
}
