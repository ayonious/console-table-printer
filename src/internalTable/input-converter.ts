import { COLOR } from '../models/common';
import { ColumnOptionsRaw } from '../models/external-table';
import { Column } from '../models/internal-table';
import { DEFAULT_ROW_ALIGNMENT } from '../utils/table-constants';

export const objIfExists = (key: string, val: any) => {
  if (!val) {
    return {};
  }

  return {
    [key]: val,
  };
};

export const rawColumnToInternalColumn = (
  column: ColumnOptionsRaw
): Column => ({
  name: column.name,
  title: column.title ?? column.name,
  ...objIfExists('color', column.color as COLOR),
  ...objIfExists('maxLen', column.maxLen),
  ...objIfExists('minLen', column.minLen),
  alignment: column.alignment || DEFAULT_ROW_ALIGNMENT,
});
