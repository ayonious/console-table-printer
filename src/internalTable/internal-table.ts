import { Dictionary, Row } from '../models/common';
import {
  ComplexOptions,
  ComputedColumn,
  RowFilterFunction,
  RowSortFunction,
} from '../models/external-table';
import { Column, TableStyleDetails } from '../models/internal-table';
import {
  DEFAULT_TABLE_STYLE,
  DEFAULT_ROW_ALIGNMENT,
  DEFAULT_ROW_FONT_COLOR,
  DEFAULT_ROW_SEPARATOR,
} from '../utils/table-constants';
import {
  createColumFromComputedColumn,
  createColumFromOnlyName,
  createRow,
  RowOptions,
} from '../utils/table-helpers';
import { rawColumnToInternalColumn } from './input-converter';
import { renderTable } from './internal-table-printer';

const DEFAULT_ROW_SORT_FUNC = () => 0;

const DEFAULT_ROW_FILTER_FUNC = () => true;

class TableInternal {
  title?: string;

  tableStyle: TableStyleDetails;

  columns: Column[];

  rows: Row[];

  filterFunction: RowFilterFunction;

  sortFunction: RowSortFunction;

  enabledColumns: string[];

  disabledColumns: string[];

  computedColumns: any[];

  rowSeparator: boolean;

  initSimple(columns: string[]) {
    this.columns = columns.map((column) => ({
      name: column,
      title: column,
      alignment: DEFAULT_ROW_ALIGNMENT,
    }));
  }

  initDetailed(options: ComplexOptions) {
    this.title = options?.title || this.title;
    this.tableStyle = options?.style || this.tableStyle;
    this.sortFunction = options?.sort || this.sortFunction;
    this.filterFunction = options?.filter || this.filterFunction;
    this.enabledColumns = options?.enabledColumns || this.enabledColumns;
    this.disabledColumns = options?.disabledColumns || this.disabledColumns;
    this.computedColumns = options?.computedColumns || this.computedColumns;
    this.columns =
      options?.columns?.map(rawColumnToInternalColumn) || this.columns;
    this.rowSeparator = options?.rowSeparator || this.rowSeparator;

    if (options.rows !== undefined) {
      this.addRows(options.rows);
    }
  }

  constructor(options?: ComplexOptions | string[]) {
    // default construction
    this.rows = [];
    this.columns = [];
    this.title = undefined;
    this.tableStyle = DEFAULT_TABLE_STYLE;
    this.filterFunction = DEFAULT_ROW_FILTER_FUNC;
    this.sortFunction = DEFAULT_ROW_SORT_FUNC;
    this.enabledColumns = [];
    this.disabledColumns = [];
    this.computedColumns = [];
    this.rowSeparator = DEFAULT_ROW_SEPARATOR;

    if (options instanceof Array) {
      this.initSimple(options);
    } else if (typeof options === 'object') {
      this.initDetailed(options);
    }
  }

  createColumnFromRow(text: Dictionary) {
    const colNames = this.columns.map((col) => col.name);
    Object.keys(text).forEach((key) => {
      if (!colNames.includes(key)) {
        this.columns.push(createColumFromOnlyName(key));
      }
    });
  }

  addColumn(textOrObj: string | ComputedColumn) {
    if (typeof textOrObj === 'string') {
      this.columns.push(createColumFromOnlyName(textOrObj));
    } else {
      this.columns.push(createColumFromComputedColumn(textOrObj));
    }
  }

  addColumns(toBeInsertedColumns: string[]) {
    toBeInsertedColumns.forEach((toBeInsertedColumn) => {
      this.addColumn(toBeInsertedColumn);
    });
  }

  addRow(text: Dictionary, options?: RowOptions) {
    this.createColumnFromRow(text);
    this.rows.push(
      createRow(
        options?.color || DEFAULT_ROW_FONT_COLOR,
        text,
        options?.separator !== undefined
          ? options?.separator
          : this.rowSeparator
      )
    );
  }

  addRows(toBeInsertedRows: Dictionary[], options?: RowOptions) {
    toBeInsertedRows.forEach((toBeInsertedRow) => {
      this.addRow(toBeInsertedRow, options);
    });
  }

  renderTable() {
    return renderTable(this);
  }
}

export default TableInternal;
