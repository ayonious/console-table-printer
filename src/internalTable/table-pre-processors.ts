/* eslint-disable no-param-reassign */
import { Row } from '../models/common';
import { ComputedColumn } from '../models/external-table';
import { Column } from '../models/internal-table';
import { findLenOfColumn } from '../utils/table-helpers';
import TableInternal from './internal-table';

const createComputedColumnsIfNecessary = (table: TableInternal) => {
  if (table.computedColumns.length) {
    table.computedColumns.forEach((computedColumn: ComputedColumn) => {
      table.addColumn(computedColumn);
      table.rows.forEach((row: Row) => {
        row.text[computedColumn.name] = computedColumn.function(row.text);
      });
    });
  }
};

const disableColumnsIfNecessary = (table: TableInternal) => {
  if (table.enabledColumns.length) {
    table.columns = table.columns.filter((col: Column) =>
      table.enabledColumns.includes(col.name)
    );
  }
};

const enableColumnsIfNecessary = (table: TableInternal) => {
  if (table.disabledColumns.length) {
    table.columns = table.columns.filter(
      (col: Column) => !table.disabledColumns.includes(col.name)
    );
  }
};

const findColumnWidth = (table: TableInternal) => {
  table.columns.forEach((column) => {
    column.length = findLenOfColumn(column, table.rows);
  });
};

export const preProcessColumns = (table: TableInternal) => {
  createComputedColumnsIfNecessary(table);
  enableColumnsIfNecessary(table);
  disableColumnsIfNecessary(table);
  findColumnWidth(table);
};

export const preProcessRows = (table: TableInternal) => {
  const newRows = table.rows
    .filter((r) => table.filterFunction(r.text))
    .sort((r1, r2) => table.sortFunction(r1.text, r2.text));
  table.rows = newRows;
};
