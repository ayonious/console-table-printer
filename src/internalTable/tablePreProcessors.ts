import { TableInternal } from './internal-table';
import { findMaxLenOfColumn, Column } from '../utils/table-helpers';

export function preProcessColumns(table: TableInternal) {
  if (table.enabledColumns.length) {
    // eslint-disable-next-line no-param-reassign
    table.columns = table.columns.filter((col: Column) =>
      table.enabledColumns.includes(col.name)
    );
  }

  if (table.disabledColumns.length) {
    // eslint-disable-next-line no-param-reassign
    table.columns = table.columns.filter(
      (col: Column) => !table.disabledColumns.includes(col.name)
    );
  }

  table.columns.forEach((column) => {
    // eslint-disable-next-line no-param-reassign
    column.max_ln = findMaxLenOfColumn(column, table.rows);
  });
}

export function preProcessRows(table: TableInternal) {
  const newrows = table.rows
    .filter((r) => table.filterFunction(r.text))
    .sort((r1, r2) => table.sortFunction(r1.text, r2.text));
  // eslint-disable-next-line no-param-reassign
  table.rows = newrows;
}
