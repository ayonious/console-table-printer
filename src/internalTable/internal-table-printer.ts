import { Row } from '../models/common';
import { Column, TableStyleDetails } from '../models/internal-table';
import ColoredConsoleLine from '../utils/colored-console-line';
import { textWithPadding } from '../utils/string-utils';
import {
  DEFAULT_COLUMN_LEN,
  DEFAULT_HEADER_ALIGNMENT,
  DEFAULT_HEADER_FONT_COLOR,
  DEFAULT_ROW_ALIGNMENT,
  DEFAULT_ROW_FONT_COLOR,
} from '../utils/table-constants';
import {
  cellText,
  createHeaderAsRow,
  createRow,
  getWidthLimitedColumnsArray,
  renderTableHorizontalBorders,
} from '../utils/table-helpers';
import TableInternal from './internal-table';
import { preProcessColumns, preProcessRows } from './table-pre-processors';

// ║ Index ║         ║        ║
const renderOneLine = (
  tableStyle: TableStyleDetails,
  columns: Column[],
  currentLineIndex: number,
  widthLimitedColumnsArray: { [key: string]: string[] },
  isHeader: boolean | undefined,
  row: Row
): string => {
  const line = new ColoredConsoleLine();
  line.addCharsWithColor(DEFAULT_ROW_FONT_COLOR, tableStyle.vertical);
  columns.forEach((column) => {
    const thisLineHasText =
      currentLineIndex < widthLimitedColumnsArray[column.name].length;

    const textForThisLine: string = thisLineHasText
      ? cellText(widthLimitedColumnsArray[column.name][currentLineIndex])
      : '';

    line.addCharsWithColor(DEFAULT_ROW_FONT_COLOR, ' ');
    line.addCharsWithColor(
      (isHeader && DEFAULT_HEADER_FONT_COLOR) || column.color || row.color,
      textWithPadding(
        textForThisLine,
        column.alignment || DEFAULT_ROW_ALIGNMENT,
        column.length || DEFAULT_COLUMN_LEN
      )
    );
    line.addCharsWithColor(DEFAULT_ROW_FONT_COLOR, ` ${tableStyle.vertical}`);
  });
  return line.renderConsole();
};

// ║ Bold  ║    text ║  value ║
// ║ Index ║         ║        ║
const renderWidthLimitedLines = (
  tableStyle: TableStyleDetails,
  columns: Column[],
  row: Row,
  isHeader?: boolean
): string[] => {
  // { col1: ['How', 'Is', 'Going'], col2: ['I am', 'Tom'],  }
  const widthLimitedColumnsArray = getWidthLimitedColumnsArray(columns, row);

  const totalLines = Object.values(widthLimitedColumnsArray).reduce(
    (a, b) => Math.max(a, b.length),
    0
  );

  const ret = [];
  for (
    let currentLineIndex = 0;
    currentLineIndex < totalLines;
    currentLineIndex += 1
  ) {
    const singleLine = renderOneLine(
      tableStyle,
      columns,
      currentLineIndex,
      widthLimitedColumnsArray,
      isHeader,
      row
    );

    ret.push(singleLine);
  }

  return ret;
};

// ║ 1     ║     I would like some red wine please ║ 10.212 ║
const renderRow = (table: TableInternal, row: Row): string[] => {
  let ret: string[] = [];
  ret = ret.concat(
    renderWidthLimitedLines(table.tableStyle, table.columns, row)
  );
  return ret;
};

/*
                  The analysis Result
 ╔═══════╦═══════════════════════════════════════╦════════╗
*/
const renderTableTitle = (table: TableInternal): string[] => {
  const ret: string[] = [];

  if (table.title === undefined) {
    return ret;
  }

  const getTableWidth = () => {
    const reducer = (accumulator: number, currentValue: number) =>
      // ║ cell ║, 2 spaces + cellTextSize + one border on the left
      accumulator + currentValue + 2 + 1;
    return table.columns
      .map((m) => m.length || DEFAULT_COLUMN_LEN)
      .reduce(reducer, 1);
  };

  const titleWithPadding = textWithPadding(
    table.title as string,
    DEFAULT_HEADER_ALIGNMENT,
    getTableWidth()
  );
  const styledText = new ColoredConsoleLine();
  styledText.addCharsWithColor(DEFAULT_HEADER_FONT_COLOR, titleWithPadding);
  //                  The analysis Result
  ret.push(styledText.renderConsole());
  return ret;
};

/*
 ╔═══════╦═══════════════════════════════════════╦════════╗
 ║ index ║                                  text ║  value ║
 ╟═══════╬═══════════════════════════════════════╬════════╢
*/
const renderTableHeaders = (table: TableInternal): string[] => {
  let ret: string[] = [];

  // ╔═══════╦═══════════════════════════════════════╦════════╗
  ret.push(
    renderTableHorizontalBorders(
      table.tableStyle.headerTop,
      table.columns.map((m: Column) => m.length || DEFAULT_COLUMN_LEN)
    )
  );

  // ║ index ║                                  text ║  value ║
  const row = createHeaderAsRow(createRow, table.columns);
  ret = ret.concat(
    renderWidthLimitedLines(table.tableStyle, table.columns, row, true)
  );

  // ╟═══════╬═══════════════════════════════════════╬════════╢
  ret.push(
    renderTableHorizontalBorders(
      table.tableStyle.headerBottom,
      table.columns.map((m) => m.length || DEFAULT_COLUMN_LEN)
    )
  );

  return ret;
};

const renderTableEnding = (table: TableInternal): string[] => {
  const ret: string[] = [];
  // ╚═══════╩═══════════════════════════════════════╩════════╝
  ret.push(
    renderTableHorizontalBorders(
      table.tableStyle.tableBottom,
      table.columns.map((m) => m.length || DEFAULT_COLUMN_LEN)
    )
  );
  return ret;
};

const renderRowSeparator = (table: TableInternal, row: Row): string[] => {
  const ret: string[] = [];
  const lastRowIndex = table.rows.length - 1;
  const currentRowIndex = table.rows.indexOf(row);

  if (currentRowIndex !== lastRowIndex && row.separator) {
    // ╟═══════╬═══════════════════════════════════════╬════════╢
    ret.push(
      renderTableHorizontalBorders(
        table.tableStyle.rowSeparator,
        table.columns.map((m) => m.length || DEFAULT_COLUMN_LEN)
      )
    );
  }
  return ret;
};

export const renderTable = (table: TableInternal): string => {
  preProcessColumns(table); // enable / disable cols, find maxLn of each col/ computed Columns
  preProcessRows(table); // sort and filter

  const ret: string[] = [];
  renderTableTitle(table).forEach((row) => ret.push(row));

  renderTableHeaders(table).forEach((row) => ret.push(row));

  table.rows.forEach((row) => {
    renderRow(table, row).forEach((row_) => ret.push(row_));
    renderRowSeparator(table, row).forEach((row_) => ret.push(row_));
  });
  renderTableEnding(table).forEach((row) => ret.push(row));
  return ret.join('\n');
};

export const renderSimpleTable = (rows: any[]) => {
  const table = new TableInternal();
  table.addRows(rows);
  return renderTable(table);
};

export const printSimpleTable = (rows: any[]) => {
  console.log(renderSimpleTable(rows));
};
