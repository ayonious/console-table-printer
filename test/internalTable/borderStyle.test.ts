import { renderTable } from '../../src/internalTable/internal-table-printer';
import { Table } from '../../index';

describe('Example: Check if borders are styled properly', () => {
  it('print a fat border table', () => {
    // Create a table
    const p = new Table({
      style: {
        /*
            Style:
            ╔══════╦═════╦══════╗
            ║ hob  ║ foo ║ mia  ║
            ╟══════╬═════╬══════╢
            ║ ball ║ fox ║ mama ║
            ╚══════╩═════╩══════╝
            */
        headerTop: {
          left: '╔',
          mid: '╦',
          right: '╗',
          other: '═',
        },
        headerBottom: {
          left: '╟',
          mid: '╬',
          right: '╢',
          other: '═',
        },
        tableBottom: {
          left: '╚',
          mid: '╩',
          right: '╝',
          other: '═',
        },
        vertical: '║',
      },
      columns: [
        { name: 'index', alignment: 'left' },
        { name: 'text', alignment: 'right' },
        { name: 'value' },
      ],
    });

    // add rows with color
    p.addRow(
      { index: 1, text: 'I would like some red wine please', value: 10.212 },
      { color: 'red' }
    );
    p.addRow(
      { index: 2, text: 'I would like some green gemuse please', value: 20.0 },
      { color: 'green' }
    );
    p.addRow(
      { index: 3, text: 'I would like some gelb bananen bitte', value: 100 },
      { color: 'yellow' }
    );

    // print
    const returned = renderTable(p.table);
    expect(returned).toMatchSnapshot();
    console.log(returned);
  });

  it('print a RED table', () => {
    // Create a table
    const p = new Table({
      style: {
        /*
            Style: (red)
            ╔══════╦═════╦══════╗
            ║ hob  ║ foo ║ mia  ║
            ╟══════╬═════╬══════╢
            ║ ball ║ fox ║ mama ║
            ╚══════╩═════╩══════╝
            */
        headerTop: {
          left: '\x1b[31m╔\x1b[0m',
          mid: '\x1b[31m╦\x1b[0m',
          right: '\x1b[31m╗\x1b[0m',
          other: '\x1b[31m═\x1b[0m',
        },
        headerBottom: {
          left: '\x1b[31m╟\x1b[0m',
          mid: '\x1b[31m╬\x1b[0m',
          right: '\x1b[31m╢\x1b[0m',
          other: '\x1b[31m═\x1b[0m',
        },
        tableBottom: {
          left: '\x1b[31m╚\x1b[0m',
          mid: '\x1b[31m╩\x1b[0m',
          right: '\x1b[31m╝\x1b[0m',
          other: '\x1b[31m═\x1b[0m',
        },
        vertical: '\x1b[31m║\x1b[0m',
      },
      columns: [
        { name: 'index', alignment: 'left' },
        { name: 'text', alignment: 'right' },
        { name: 'value' },
      ],
    });

    // add rows with color
    p.addRow(
      { index: 1, text: 'I would like some red wine please', value: 10.212 },
      { color: 'red' }
    );
    p.addRow(
      { index: 2, text: 'I would like some green gemuse please', value: 20.0 },
      { color: 'green' }
    );
    p.addRow(
      { index: 3, text: 'I would like some gelb bananen bitte', value: 100 },
      { color: 'yellow' }
    );

    // print
    const returned = renderTable(p.table);
    expect(returned).toMatchSnapshot();
    console.log(returned);
  });
});
