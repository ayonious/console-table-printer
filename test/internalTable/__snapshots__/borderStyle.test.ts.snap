// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example: Check if borders are styled properly print a RED table 1`] = `
"[31m╔[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m╦[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m╦[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m╗[0m
[31m║[0m[37m [0m[01mindex[0m [31m║[0m[37m [0m[01m                                 text[0m [31m║[0m[37m [0m[01m value[0m [31m║[0m
[31m╟[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m╬[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m╬[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m╢[0m
[31m║[0m[37m [0m[31m1    [0m [31m║[0m[37m [0m[31m    I would like some red wine please[0m [31m║[0m[37m [0m[31m10.212[0m [31m║[0m
[31m║[0m[37m [0m[32m2    [0m [31m║[0m[37m [0m[32mI would like some green gemuse please[0m [31m║[0m[37m [0m[32m    20[0m [31m║[0m
[31m║[0m[37m [0m[33m3    [0m [31m║[0m[37m [0m[33m I would like some gelb bananen bitte[0m [31m║[0m[37m [0m[33m   100[0m [31m║[0m
[31m╚[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m╩[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m╩[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m═[0m[31m╝[0m"
`;

exports[`Example: Check if borders are styled properly print a fat border table 1`] = `
"╔═══════╦═══════════════════════════════════════╦════════╗
║[37m [0m[01mindex[0m ║[37m [0m[01m                                 text[0m ║[37m [0m[01m value[0m ║
╟═══════╬═══════════════════════════════════════╬════════╢
║[37m [0m[31m1    [0m ║[37m [0m[31m    I would like some red wine please[0m ║[37m [0m[31m10.212[0m ║
║[37m [0m[32m2    [0m ║[37m [0m[32mI would like some green gemuse please[0m ║[37m [0m[32m    20[0m ║
║[37m [0m[33m3    [0m ║[37m [0m[33m I would like some gelb bananen bitte[0m ║[37m [0m[33m   100[0m ║
╚═══════╩═══════════════════════════════════════╩════════╝"
`;
